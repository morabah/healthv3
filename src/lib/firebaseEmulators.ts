import { getApp } from 'firebase/app';
import { connectAuthEmulator, getAuth } from 'firebase/auth';
import { connectFirestoreEmulator, getFirestore } from 'firebase/firestore';
import { connectFunctionsEmulator, getFunctions } from 'firebase/functions';
import { connectStorageEmulator, getStorage } from 'firebase/storage';

/**
 * Connect to Firebase emulators in development and test environments
 */
export const connectToEmulators = () => {
  try {
    // Only connect to emulators in development or test environment
    if (process.env.NODE_ENV === 'production') {
      return;
    }

    const app = getApp();
    
    // Auth emulator
    const auth = getAuth(app);
    connectAuthEmulator(auth, 'http://localhost:9099', { disableWarnings: true });
    
    // Firestore emulator
    const firestore = getFirestore(app);
    connectFirestoreEmulator(firestore, 'localhost', 8080);
    
    // Functions emulator
    const functions = getFunctions(app);
    connectFunctionsEmulator(functions, 'localhost', 5001);
    
    // Storage emulator
    const storage = getStorage(app);
    connectStorageEmulator(storage, 'localhost', 9199);
    
    console.log('Connected to Firebase emulators');
  } catch (error) {
    console.error('Error connecting to Firebase emulators:', error);
  }
};

/**
 * Helper function to check if emulators are running
 */
export const areEmulatorsRunning = async (): Promise<boolean> => {
  try {
    // Try to connect to the Emulator UI
    const response = await fetch('http://localhost:4000');
    return response.status === 200;
  } catch (error) {
    return false;
  }
};
