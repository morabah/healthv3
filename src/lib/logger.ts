/**
 * Logging utility for the Health Appointment System
 * Provides standardized logging functions with timestamps and log levels
 */

/**
 * Returns current ISO timestamp for consistent log formatting
 */
const getTimestamp = (): string => {
  return new Date().toISOString();
};

/**
 * Type definition for log parameters to ensure type safety
 */
export type LogParams = {
  message: string;
  data?: unknown;
  context?: string;
};

/**
 * Formats log message with timestamp, level, context, and optional data
 */
const formatLogMessage = (
  level: string,
  { message, data, context }: LogParams
): string => {
  const timestamp = getTimestamp();
  const contextStr = context ? `[${context}] ` : '';
  const baseMessage = `[${timestamp}] [${level}] ${contextStr}${message}`;
  
  if (data !== undefined) {
    return `${baseMessage} ${JSON.stringify(data, null, 2)}`;
  }
  
  return baseMessage;
};

/**
 * Logs informational messages
 * @param params Log parameters including message, optional data, and context
 */
export const logInfo = (params: LogParams): void => {
  console.log(formatLogMessage('INFO', params));
};

/**
 * Logs warning messages
 * @param params Log parameters including message, optional data, and context
 */
export const logWarn = (params: LogParams): void => {
  console.warn(formatLogMessage('WARN', params));
};

/**
 * Logs error messages
 * @param params Log parameters including message, optional data, and context
 */
export const logError = (params: LogParams): void => {
  console.error(formatLogMessage('ERROR', params));
};

/**
 * Logs debug messages only when NEXT_PUBLIC_LOG_LEVEL is set to 'debug'
 * @param params Log parameters including message, optional data, and context
 */
export const logDebug = (params: LogParams): void => {
  if (typeof process !== 'undefined' && 
      process.env.NEXT_PUBLIC_LOG_LEVEL === 'debug') {
    console.debug(formatLogMessage('DEBUG', params));
  }
};

/**
 * Default export for convenient importing
 */
export default {
  logInfo,
  logWarn,
  logError,
  logDebug,
};
