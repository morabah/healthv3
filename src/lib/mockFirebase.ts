/**
 * Mock Firebase implementation for local testing
 * This allows testing authentication flows without needing Firebase emulators
 */

import { UserType, VerificationStatus } from '@/types/enums';

// Mock user storage
const mockUsers: Record<string, any> = {};
const mockUserProfiles: Record<string, any> = {};
const mockPatients: Record<string, any> = {};
const mockDoctors: Record<string, any> = {};

// Mock current user
let currentUser: any = null;

// Mock Firebase Auth
export const mockAuth = {
  currentUser: null as any,
  
  // Sign in with email and password
  signInWithEmailAndPassword: async (email: string, password: string) => {
    const user = Object.values(mockUsers).find(
      (u: any) => u.email === email && u.password === password
    );
    
    if (!user) {
      throw new Error('auth/user-not-found');
    }
    
    currentUser = {
      uid: user.uid,
      email: user.email,
      emailVerified: user.emailVerified,
      reload: async () => {
        // Simulate reload by updating from mock storage
        const updatedUser = mockUsers[user.uid];
        if (updatedUser) {
          currentUser.emailVerified = updatedUser.emailVerified;
        }
      }
    };
    
    mockAuth.currentUser = currentUser;
    
    // Trigger onAuthStateChanged callbacks
    mockAuth._listeners.forEach((callback: any) => {
      callback(currentUser);
    });
    
    return { user: currentUser };
  },
  
  // Create user with email and password
  createUserWithEmailAndPassword: async (email: string, password: string) => {
    // Check if user already exists
    const existingUser = Object.values(mockUsers).find(
      (u: any) => u.email === email
    );
    
    if (existingUser) {
      throw new Error('auth/email-already-exists');
    }
    
    // Create new user
    const uid = `user_${Date.now()}`;
    const newUser = {
      uid,
      email,
      password, // Note: In a real system, this would be hashed
      emailVerified: false,
      reload: async () => {
        // Simulate reload by updating from mock storage
        const updatedUser = mockUsers[uid];
        if (updatedUser) {
          currentUser.emailVerified = updatedUser.emailVerified;
        }
      }
    };
    
    mockUsers[uid] = newUser;
    currentUser = {
      uid: newUser.uid,
      email: newUser.email,
      emailVerified: newUser.emailVerified,
      reload: newUser.reload
    };
    
    mockAuth.currentUser = currentUser;
    
    // Trigger onAuthStateChanged callbacks
    mockAuth._listeners.forEach((callback: any) => {
      callback(currentUser);
    });
    
    return { user: currentUser };
  },
  
  // Sign out
  signOut: async () => {
    currentUser = null;
    mockAuth.currentUser = null;
    
    // Trigger onAuthStateChanged callbacks
    mockAuth._listeners.forEach((callback: any) => {
      callback(null);
    });
    
    return Promise.resolve();
  },
  
  // Send email verification
  sendEmailVerification: async (user: any) => {
    // In a real system, this would send an email
    console.log(`Sending verification email to ${user.email}`);
    return Promise.resolve();
  },
  
  // Verify email (mock function for testing)
  verifyEmail: async (uid: string) => {
    if (mockUsers[uid]) {
      mockUsers[uid].emailVerified = true;
      
      if (currentUser && currentUser.uid === uid) {
        currentUser.emailVerified = true;
      }
    }
    
    return Promise.resolve();
  },
  
  // Auth state change listener
  _listeners: [] as any[],
  onAuthStateChanged: (callback: any) => {
    mockAuth._listeners.push(callback);
    
    // Immediately call with current state
    callback(currentUser);
    
    // Return unsubscribe function
    return () => {
      mockAuth._listeners = mockAuth._listeners.filter(
        (cb: any) => cb !== callback
      );
    };
  }
};

// Mock Firebase Functions
export const mockFunctions = {
  httpsCallable: (name: string) => {
    // Mock registerUser function
    if (name === 'registerUser') {
      return async (data: any) => {
        const { 
          email, 
          password, 
          userType, 
          firstName, 
          lastName, 
          phone, 
          patientData, 
          doctorData 
        } = data;
        
        try {
          // Create user in mock Auth
          const { user } = await mockAuth.createUserWithEmailAndPassword(email, password);
          
          // Create user profile
          const userProfile = {
            userId: user.uid,
            email,
            firstName,
            lastName,
            phone: phone || null,
            userType,
            createdAt: new Date().toISOString(),
            updatedAt: new Date().toISOString(),
            emailVerified: false,
            phoneVerified: false,
            verificationStatus: userType === UserType.DOCTOR 
              ? VerificationStatus.PENDING 
              : VerificationStatus.VERIFIED,
          };
          
          mockUserProfiles[user.uid] = userProfile;
          
          // Create type-specific profile
          if (userType === UserType.PATIENT && patientData) {
            const patientProfile = {
              userId: user.uid,
              dateOfBirth: patientData.dateOfBirth,
              gender: patientData.gender,
              bloodType: patientData.bloodType || null,
              createdAt: new Date().toISOString(),
              updatedAt: new Date().toISOString(),
            };
            
            mockPatients[user.uid] = patientProfile;
          } else if (userType === UserType.DOCTOR && doctorData) {
            const doctorProfile = {
              userId: user.uid,
              specialty: doctorData.specialty,
              licenseNumber: doctorData.licenseNumber,
              yearsOfExperience: doctorData.yearsOfExperience,
              location: doctorData.location,
              languages: doctorData.languages,
              consultationFee: doctorData.consultationFee,
              profilePictureUrl: doctorData.profilePictureUrl,
              licenseDocumentUrl: doctorData.licenseDocumentUrl,
              verificationStatus: VerificationStatus.PENDING,
              createdAt: new Date().toISOString(),
              updatedAt: new Date().toISOString(),
            };
            
            mockDoctors[user.uid] = doctorProfile;
          }
          
          console.log('Mock registration successful:', {
            user,
            userProfile,
            patientData: patientData ? mockPatients[user.uid] : null,
            doctorData: doctorData ? mockDoctors[user.uid] : null
          });
          
          // Return data in the format expected by the Cloud Functions SDK
          return {
            data: {
              success: true,
              userId: user.uid,
              message: 'User registered successfully',
            }
          };
        } catch (error: any) {
          console.error('Mock registration error:', error);
          
          // Format error to match Firebase Functions error format
          throw {
            code: error.message || 'unknown-error',
            message: error.message || 'An unknown error occurred during registration'
          };
        }
      };
    }
    
    // Default mock function
    return async () => {
      return { data: { success: true } };
    };
  }
};

// Mock Firebase Firestore
export const mockFirestore = {
  // Collection references
  collection: (collectionPath: string) => {
    return {
      doc: (docId: string) => {
        return {
          get: async () => {
            let data: any = null;
            
            // Determine which mock storage to use based on collection path
            if (collectionPath === 'users' && mockUserProfiles[docId]) {
              data = mockUserProfiles[docId];
            } else if (collectionPath === 'patients' && mockPatients[docId]) {
              data = mockPatients[docId];
            } else if (collectionPath === 'doctors' && mockDoctors[docId]) {
              data = mockDoctors[docId];
            }
            
            return {
              exists: !!data,
              data: () => data,
              id: docId
            };
          },
          set: async (data: any) => {
            // Determine which mock storage to use based on collection path
            if (collectionPath === 'users') {
              mockUserProfiles[docId] = { ...data };
            } else if (collectionPath === 'patients') {
              mockPatients[docId] = { ...data };
            } else if (collectionPath === 'doctors') {
              mockDoctors[docId] = { ...data };
            }
            return Promise.resolve();
          }
        };
      }
    };
  },
  
  // Get a document reference
  doc: (path: string) => {
    const parts = path.split('/');
    const collectionName = parts[0];
    const docId = parts[1];
    
    return {
      get: async () => {
        let data: any = null;
        
        // Determine which mock storage to use based on collection path
        if (collectionName === 'users' && mockUserProfiles[docId]) {
          data = mockUserProfiles[docId];
        } else if (collectionName === 'patients' && mockPatients[docId]) {
          data = mockPatients[docId];
        } else if (collectionName === 'doctors' && mockDoctors[docId]) {
          data = mockDoctors[docId];
        }
        
        return {
          exists: !!data,
          data: () => data,
          id: docId
        };
      }
    };
  },
  
  // Get document data
  getDoc: async (docRef: any) => {
    return await docRef.get();
  },
  
  // Legacy methods for backward compatibility
  getUserProfile: async (uid: string) => {
    return mockUserProfiles[uid] || null;
  },
  
  getPatientProfile: async (uid: string) => {
    return mockPatients[uid] || null;
  },
  
  getDoctorProfile: async (uid: string) => {
    return mockDoctors[uid] || null;
  }
};

// Mock Firebase Storage
export const mockStorage = {
  uploadBytes: async () => {
    return Promise.resolve();
  },
  getDownloadURL: async () => {
    return 'https://example.com/mock-image-url.jpg';
  }
};

// Debug function to view all mock data
export const debugMockData = () => {
  return {
    users: mockUsers,
    userProfiles: mockUserProfiles,
    patients: mockPatients,
    doctors: mockDoctors,
    currentUser
  };
};
