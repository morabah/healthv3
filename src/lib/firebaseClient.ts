/**
 * Firebase Client SDK Initialization
 * This module initializes and exports Firebase client services for use in the frontend
 */

import { initializeApp, getApps, FirebaseApp } from 'firebase/app';
import { getAuth, Auth } from 'firebase/auth';
import { getFirestore, Firestore } from 'firebase/firestore';
import { getStorage, FirebaseStorage } from 'firebase/storage';
import { getFunctions, Functions } from 'firebase/functions';
import { logInfo, logError } from './logger';
import { connectToEmulators } from './firebaseEmulators';

// Firebase configuration from environment variables
const firebaseConfig = {
  apiKey: process.env.NEXT_PUBLIC_FIREBASE_API_KEY,
  authDomain: process.env.NEXT_PUBLIC_FIREBASE_AUTH_DOMAIN,
  projectId: process.env.NEXT_PUBLIC_FIREBASE_PROJECT_ID,
  storageBucket: process.env.NEXT_PUBLIC_FIREBASE_STORAGE_BUCKET,
  messagingSenderId: process.env.NEXT_PUBLIC_FIREBASE_MESSAGING_SENDER_ID,
  appId: process.env.NEXT_PUBLIC_FIREBASE_APP_ID,
  measurementId: process.env.NEXT_PUBLIC_FIREBASE_MEASUREMENT_ID,
};

// Singleton instances
let firebaseApp: FirebaseApp;
let firebaseAuth: Auth;
let firebaseFirestore: Firestore;
let firebaseStorage: FirebaseStorage;
let firebaseFunctions: Functions;

/**
 * Initialize Firebase if it hasn't been initialized already
 * @returns The Firebase app instance
 */
export const initializeFirebase = (): FirebaseApp => {
  try {
    if (!firebaseApp && getApps().length === 0) {
      logInfo({
        message: 'Initializing Firebase client SDK',
        context: 'firebaseClient',
      });
      
      // Check if required environment variables are set
      if (!process.env.NEXT_PUBLIC_FIREBASE_API_KEY || !process.env.NEXT_PUBLIC_FIREBASE_PROJECT_ID) {
        throw new Error('Firebase environment variables are not properly configured');
      }
      
      firebaseApp = initializeApp(firebaseConfig);
      
      // Connect to emulators in development and test environments
      if (process.env.NODE_ENV !== 'production') {
        connectToEmulators();
      }
      
      logInfo({
        message: 'Firebase client SDK initialized successfully',
        context: 'firebaseClient',
        data: { projectId: process.env.NEXT_PUBLIC_FIREBASE_PROJECT_ID }
      });
    } else if (!firebaseApp) {
      firebaseApp = getApps()[0];
    }
    
    return firebaseApp;
  } catch (error) {
    logError({
      message: 'Failed to initialize Firebase client SDK',
      context: 'firebaseClient',
      data: { error }
    });
    
    throw error;
  }
};

/**
 * Get Firebase Auth instance
 * @returns Firebase Auth instance
 */
export const getFirebaseAuth = (): Auth => {
  if (!firebaseAuth) {
    const app = initializeFirebase();
    firebaseAuth = getAuth(app);
    logInfo({
      message: 'Firebase Auth initialized',
      context: 'firebaseClient'
    });
  }
  return firebaseAuth;
};

/**
 * Get Firebase Firestore instance
 * @returns Firebase Firestore instance
 */
export const getFirebaseFirestore = (): Firestore => {
  if (!firebaseFirestore) {
    const app = initializeFirebase();
    firebaseFirestore = getFirestore(app);
    logInfo({
      message: 'Firebase Firestore initialized',
      context: 'firebaseClient'
    });
  }
  return firebaseFirestore;
};

/**
 * Get Firebase Storage instance
 * @returns Firebase Storage instance
 */
export const getFirebaseStorage = (): FirebaseStorage => {
  if (!firebaseStorage) {
    const app = initializeFirebase();
    firebaseStorage = getStorage(app);
    logInfo({
      message: 'Firebase Storage initialized',
      context: 'firebaseClient'
    });
  }
  return firebaseStorage;
};

/**
 * Get Firebase Functions instance
 * @returns Firebase Functions instance
 */
export const getFirebaseFunctions = (): Functions => {
  if (!firebaseFunctions) {
    const app = initializeFirebase();
    firebaseFunctions = getFunctions(app);
    logInfo({
      message: 'Firebase Functions initialized',
      context: 'firebaseClient'
    });
  }
  return firebaseFunctions;
};

// Initialize Firebase on module load in browser environments
if (typeof window !== 'undefined') {
  try {
    initializeFirebase();
    logInfo({
      message: 'Firebase auto-initialized in browser environment',
      context: 'firebaseClient'
    });
  } catch (error) {
    logError({
      message: 'Error during Firebase auto-initialization',
      context: 'firebaseClient',
      data: { error }
    });
  }
}

// Default export for convenient importing
export default {
  initializeFirebase,
  getFirebaseAuth,
  getFirebaseFirestore,
  getFirebaseStorage,
  getFirebaseFunctions,
};
