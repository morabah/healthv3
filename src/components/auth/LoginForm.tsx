/**
 * Login Form Component
 * Handles user authentication with email and password
 */
import React, { useState } from 'react';
import { getAuth, signInWithEmailAndPassword } from 'firebase/auth';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faEnvelope, faLock, faSpinner } from '@fortawesome/free-solid-svg-icons';
import Button from '../ui/Button';
import { logInfo, logError } from '@/lib/logger';
import { trackPerformance } from '@/lib/performance';

interface FormData {
  email: string;
  password: string;
}

interface FormErrors {
  email?: string;
  password?: string;
  general?: string;
}

const LoginForm: React.FC = () => {
  const [formData, setFormData] = useState<FormData>({
    email: '',
    password: '',
  });
  const [errors, setErrors] = useState<FormErrors>({});
  const [isLoading, setIsLoading] = useState(false);

  const validateForm = (): boolean => {
    const newErrors: FormErrors = {};
    let isValid = true;

    // Email validation
    if (!formData.email) {
      newErrors.email = 'Email is required';
      isValid = false;
    } else if (!/\S+@\S+\.\S+/.test(formData.email)) {
      newErrors.email = 'Please enter a valid email address';
      isValid = false;
    }

    // Password validation
    if (!formData.password) {
      newErrors.password = 'Password is required';
      isValid = false;
    }

    setErrors(newErrors);
    return isValid;
  };

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { name, value } = e.target;
    setFormData((prev) => ({
      ...prev,
      [name]: value,
    }));
  };

  const handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    if (!validateForm()) {
      return;
    }

    setIsLoading(true);
    setErrors({});

    try {
      const performanceTracker = trackPerformance('login', 'auth');
      
      logInfo({
        message: 'Login attempt',
        context: 'LoginForm',
        data: { email: formData.email },
      });

      const auth = getAuth();
      await signInWithEmailAndPassword(auth, formData.email, formData.password);
      
      performanceTracker.stop();
      
      logInfo({
        message: 'Login successful',
        context: 'LoginForm',
        data: { email: formData.email },
      });
    } catch (error: any) {
      logError({
        message: 'Login failed',
        context: 'LoginForm',
        data: { 
          email: formData.email,
          errorCode: error.code,
          errorMessage: error.message 
        },
      });

      // Handle specific Firebase auth errors
      let errorMessage = 'Failed to sign in. Please try again.';
      
      if (error.code === 'auth/user-not-found') {
        errorMessage = 'No account found with this email address';
      } else if (error.code === 'auth/wrong-password') {
        errorMessage = 'Invalid password';
      } else if (error.code === 'auth/too-many-requests') {
        errorMessage = 'Too many failed login attempts. Please try again later';
      } else if (error.code === 'auth/invalid-credential') {
        errorMessage = 'Invalid email or password';
      }
      
      setErrors({ general: errorMessage });
    } finally {
      setIsLoading(false);
    }
  };

  return (
    <div className="card">
      <div className="card-header">
        <h2 className="text-xl font-semibold">Sign In</h2>
      </div>
      
      <form onSubmit={handleSubmit} className="card-body">
        {errors.general && (
          <div className="error-text mb-4">{errors.general}</div>
        )}
        
        <div className="mb-4">
          <label htmlFor="email" className="block mb-2 font-medium">
            Email
          </label>
          <div className="relative">
            <span className="absolute left-3 top-3 text-gray-500">
              <FontAwesomeIcon icon={faEnvelope} />
            </span>
            <input
              id="email"
              name="email"
              type="email"
              className="form-input pl-10"
              placeholder="you@example.com"
              value={formData.email}
              onChange={handleChange}
              disabled={isLoading}
              aria-label="Email"
            />
          </div>
          {errors.email && <div className="error-text">{errors.email}</div>}
        </div>
        
        <div className="mb-6">
          <label htmlFor="password" className="block mb-2 font-medium">
            Password
          </label>
          <div className="relative">
            <span className="absolute left-3 top-3 text-gray-500">
              <FontAwesomeIcon icon={faLock} />
            </span>
            <input
              id="password"
              name="password"
              type="password"
              className="form-input pl-10"
              placeholder="••••••••"
              value={formData.password}
              onChange={handleChange}
              disabled={isLoading}
              aria-label="Password"
            />
          </div>
          {errors.password && <div className="error-text">{errors.password}</div>}
        </div>
        
        <div className="flex items-center justify-between">
          <Button
            type="submit"
            loading={isLoading}
            disabled={isLoading}
            icon={isLoading ? faSpinner : undefined}
          >
            Sign In
          </Button>
          
          <a href="/auth/forgot-password" className="text-blue-600 hover:text-blue-800">
            Forgot password?
          </a>
        </div>
      </form>
      
      <div className="card-footer">
        <p className="text-center">
          Don't have an account?{' '}
          <a href="/auth/register" className="text-blue-600 hover:text-blue-800">
            Create one
          </a>
        </p>
      </div>
    </div>
  );
};

export default LoginForm;
