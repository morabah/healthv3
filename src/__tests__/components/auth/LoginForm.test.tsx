/**
 * LoginForm Component Tests
 * Tests the login form component with Firebase auth emulator
 */
import React from 'react';
import { render, waitFor, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import LoginForm from '@/components/auth/LoginForm';
import { signInWithEmailAndPassword } from 'firebase/auth';
import { logInfo, logError } from '@/lib/logger';

// Mock Firebase auth
jest.mock('firebase/auth', () => ({
  getAuth: jest.fn(() => ({})),
  signInWithEmailAndPassword: jest.fn(),
}));

// Mock logger and performance utilities
jest.mock('@/lib/logger', () => ({
  logInfo: jest.fn(),
  logError: jest.fn(),
}));

jest.mock('@/lib/performance', () => ({
  trackPerformance: jest.fn(() => ({
    stop: jest.fn(),
  })),
}));

describe('LoginForm Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders login form correctly', () => {
    const { getByLabelText, getByRole } = render(<LoginForm />);
    
    expect(getByLabelText(/email/i)).toBeInTheDocument();
    expect(getByLabelText(/password/i)).toBeInTheDocument();
    expect(getByRole('button', { name: /sign in/i })).toBeInTheDocument();
    expect(screen.getByText(/forgot password/i)).toBeInTheDocument();
    expect(screen.getByText(/don't have an account/i)).toBeInTheDocument();
  });

  it('submits the form with valid data', async () => {
    const mockSignIn = signInWithEmailAndPassword as jest.Mock;
    mockSignIn.mockResolvedValueOnce({ user: { email: 'test@example.com' } });
    
    const { getByLabelText, getByRole } = render(<LoginForm />);
    const user = userEvent.setup();
    
    // Fill in valid data
    await user.type(getByLabelText(/email/i), 'test@example.com');
    await user.type(getByLabelText(/password/i), 'Password123!');
    
    // Submit the form
    await user.click(getByRole('button', { name: /sign in/i }));
    
    // Check if Firebase auth was called with correct parameters
    await waitFor(() => {
      expect(mockSignIn).toHaveBeenCalledWith(
        expect.anything(),
        'test@example.com',
        'Password123!'
      );
    });
    
    // Verify logging
    expect(logInfo).toHaveBeenCalledWith(
      expect.objectContaining({
        message: 'Login attempt',
        context: 'LoginForm',
        data: expect.objectContaining({ email: 'test@example.com' })
      })
    );
    
    expect(logInfo).toHaveBeenCalledWith(
      expect.objectContaining({
        message: 'Login successful',
        context: 'LoginForm'
      })
    );
  });
  
  it('shows validation errors for empty fields', async () => {
    const { getByRole, getByText } = render(<LoginForm />);
    const user = userEvent.setup();
    
    // Submit form without filling any fields
    await user.click(getByRole('button', { name: /sign in/i }));
    
    // Check for validation error messages
    expect(getByText('Email is required')).toBeInTheDocument();
    expect(getByText('Password is required')).toBeInTheDocument();
    
    // Verify Firebase auth was not called
    expect(signInWithEmailAndPassword).not.toHaveBeenCalled();
  });
  
  it('validates email format', async () => {
    const { getByLabelText, getByRole, getByText } = render(<LoginForm />);
    const user = userEvent.setup();
    
    // Fill in invalid email format
    await user.type(getByLabelText(/email/i), 'invalid-email');
    await user.type(getByLabelText(/password/i), 'Password123!');
    
    // Submit the form
    await user.click(getByRole('button', { name: /sign in/i }));
    
    // Check for validation error message
    expect(getByText('Please enter a valid email address')).toBeInTheDocument();
    
    // Verify Firebase auth was not called
    expect(signInWithEmailAndPassword).not.toHaveBeenCalled();
  });
  
  it('handles Firebase auth "user not found" error', async () => {
    const mockSignIn = signInWithEmailAndPassword as jest.Mock;
    mockSignIn.mockRejectedValueOnce({ code: 'auth/user-not-found', message: 'User not found' });
    
    const { getByLabelText, getByRole, findByText } = render(<LoginForm />);
    const user = userEvent.setup();
    
    // Fill in valid data
    await user.type(getByLabelText(/email/i), 'nonexistent@example.com');
    await user.type(getByLabelText(/password/i), 'Password123!');
    
    // Submit the form
    await user.click(getByRole('button', { name: /sign in/i }));
    
    // Check for error message
    const errorMessage = await findByText('No account found with this email address');
    expect(errorMessage).toBeInTheDocument();
    
    // Verify logging
    expect(logError).toHaveBeenCalledWith(
      expect.objectContaining({
        message: 'Login failed',
        context: 'LoginForm',
        data: expect.objectContaining({ 
          errorCode: 'auth/user-not-found'
        })
      })
    );
  });
  
  it('handles Firebase auth "wrong password" error', async () => {
    const mockSignIn = signInWithEmailAndPassword as jest.Mock;
    mockSignIn.mockRejectedValueOnce({ code: 'auth/wrong-password', message: 'Wrong password' });
    
    const { getByLabelText, getByRole, findByText } = render(<LoginForm />);
    const user = userEvent.setup();
    
    // Fill in valid data with wrong password
    await user.type(getByLabelText(/email/i), 'test@example.com');
    await user.type(getByLabelText(/password/i), 'WrongPassword123!');
    
    // Submit the form
    await user.click(getByRole('button', { name: /sign in/i }));
    
    // Check for error message
    const errorMessage = await findByText('Invalid password');
    expect(errorMessage).toBeInTheDocument();
  });
  
  it('handles Firebase auth "too many requests" error', async () => {
    const mockSignIn = signInWithEmailAndPassword as jest.Mock;
    mockSignIn.mockRejectedValueOnce({ code: 'auth/too-many-requests', message: 'Too many requests' });
    
    const { getByLabelText, getByRole, findByText } = render(<LoginForm />);
    const user = userEvent.setup();
    
    // Fill in valid data
    await user.type(getByLabelText(/email/i), 'test@example.com');
    await user.type(getByLabelText(/password/i), 'Password123!');
    
    // Submit the form
    await user.click(getByRole('button', { name: /sign in/i }));
    
    // Check for error message
    const errorMessage = await findByText('Too many failed login attempts. Please try again later');
    expect(errorMessage).toBeInTheDocument();
  });
  
  it('disables form inputs and shows loading state during submission', async () => {
    // Use a promise that we can resolve manually to control the timing
    let resolvePromise: (value: any) => void;
    const authPromise = new Promise((resolve) => {
      resolvePromise = resolve;
    });
    
    const mockSignIn = signInWithEmailAndPassword as jest.Mock;
    mockSignIn.mockReturnValueOnce(authPromise);
    
    const { getByLabelText, getByRole } = render(<LoginForm />);
    const user = userEvent.setup();
    
    // Fill in valid data
    await user.type(getByLabelText(/email/i), 'test@example.com');
    await user.type(getByLabelText(/password/i), 'Password123!');
    
    // Submit the form
    await user.click(getByRole('button', { name: /sign in/i }));
    
    // Check loading state
    const emailInput = getByLabelText(/email/i) as HTMLInputElement;
    const passwordInput = getByLabelText(/password/i) as HTMLInputElement;
    const submitButton = getByRole('button', { name: /sign in/i });
    
    expect(emailInput.disabled).toBe(true);
    expect(passwordInput.disabled).toBe(true);
    expect(submitButton).toBeDisabled();
    
    // Resolve the auth promise
    resolvePromise!({ user: { email: 'test@example.com' } });
    
    // Wait for the form to return to normal state
    await waitFor(() => {
      expect(emailInput.disabled).toBe(false);
      expect(passwordInput.disabled).toBe(false);
      expect(submitButton).not.toBeDisabled();
    });
  });
  
  it('matches snapshot', () => {
    const { container } = render(<LoginForm />);
    expect(container).toMatchSnapshot();
  });
});
