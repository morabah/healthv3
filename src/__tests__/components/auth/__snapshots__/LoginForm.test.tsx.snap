// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm Component matches snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <div
      class="card-header"
    >
      <h2
        class="text-xl font-semibold"
      >
        Sign In
      </h2>
    </div>
    <form
      class="card-body"
    >
      <div
        class="mb-4"
      >
        <label
          class="block mb-2 font-medium"
          for="email"
        >
          Email
        </label>
        <div
          class="relative"
        >
          <span
            class="absolute left-3 top-3 text-gray-500"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-envelope "
              data-icon="envelope"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M48 64C21.5 64 0 85.5 0 112c0 15.1 7.1 29.3 19.2 38.4L236.8 313.6c11.4 8.5 27 8.5 38.4 0L492.8 150.4c12.1-9.1 19.2-23.3 19.2-38.4c0-26.5-21.5-48-48-48L48 64zM0 176L0 384c0 35.3 28.7 64 64 64l384 0c35.3 0 64-28.7 64-64l0-208L294.4 339.2c-22.8 17.1-54 17.1-76.8 0L0 176z"
                fill="currentColor"
              />
            </svg>
          </span>
          <input
            aria-label="Email"
            class="form-input pl-10"
            id="email"
            name="email"
            placeholder="you@example.com"
            type="email"
            value=""
          />
        </div>
      </div>
      <div
        class="mb-6"
      >
        <label
          class="block mb-2 font-medium"
          for="password"
        >
          Password
        </label>
        <div
          class="relative"
        >
          <span
            class="absolute left-3 top-3 text-gray-500"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-lock "
              data-icon="lock"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M144 144l0 48 160 0 0-48c0-44.2-35.8-80-80-80s-80 35.8-80 80zM80 192l0-48C80 64.5 144.5 0 224 0s144 64.5 144 144l0 48 16 0c35.3 0 64 28.7 64 64l0 192c0 35.3-28.7 64-64 64L64 512c-35.3 0-64-28.7-64-64L0 256c0-35.3 28.7-64 64-64l16 0z"
                fill="currentColor"
              />
            </svg>
          </span>
          <input
            aria-label="Password"
            class="form-input pl-10"
            id="password"
            name="password"
            placeholder="••••••••"
            type="password"
            value=""
          />
        </div>
      </div>
      <div
        class="flex items-center justify-between"
      >
        <button
          class="
    inline-flex items-center justify-center font-medium transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2
    text-base py-2 px-4
    bg-blue-600 hover:bg-blue-700 text-white focus:ring-blue-500 bg-blue
    
    
    rounded-md
    
  "
          type="submit"
        >
          Sign In
        </button>
        <a
          class="text-blue-600 hover:text-blue-800"
          href="/auth/forgot-password"
        >
          Forgot password?
        </a>
      </div>
    </form>
    <div
      class="card-footer"
    >
      <p
        class="text-center"
      >
        Don't have an account?
         
        <a
          class="text-blue-600 hover:text-blue-800"
          href="/auth/register"
        >
          Create one
        </a>
      </p>
    </div>
  </div>
</div>
`;
